src/lifecycle/created.js

```diff
@@ -46,46 +49,79 @@ function callCreatedOnDescendants (elem, id) {
-  var template = opts.template || function () {};
+  var template = fnOr(opts.template);
 
   /* jshint expr: true */
   return function () {
+    var initEvents, initProps;
     var info = data(this, opts.id);
     var isNative = this.createdCallback;
 
-    if (info.created) {
-      return;
-    }
-
+    // This ensures that this component cannot be triggered more than once on
+    // this element. This is also for native so that we can force-init an
+    // element and when the native callback is fired it can just be a no-op.
+    if (info.created) return;
     info.created = true;
-    isNative || prototype.call(this);
+
+    // Ensure that attribute changes are propagated synchronously.
     isNative || patchAttributeMethods(this);
-    events.call(this);
+
+    // The inheriting of the prototype is the first thing to happen in native.
+    isNative || prototype.call(this);
+
+    // Bind events so we can catch them at the earliest point in the lifecycle.
+    // Any events that happen between here and when `initEvents()` is called
+    // are queued up and not invoked until `initEvents()` is actually called.
+    // This allows us to ensure that everything is set up before we actually
+    // handle the events.
+    initEvents = events.call(this);
+
+    // The properties function returns a function that can be called to
+    // initialise them on the element when appropriate. We bind property
+    // handlers before calling any of the lifecycle handlers, but don't
+    // actually initialise their values until after lifecycle callbacks have
+    // been invoked.
+    initProps = properties.call(this);
+
+    // Apply the template to the element. In native, this may cause whatever
+    // descendants are in the template (including existing innerHTML) to be
+    // upgraded synchronously.
     template.call(this);
-    properties.call(this);
+
+    // In both polyfill and native we force init all descendant components so
+    // that we can ensure that created() has been called on all descendants
+    // by the time it's called on the host.
+    callCreatedOnDescendants(this, opts.id);
+
+    // Call created() on the host. This may be called at any time, so no one
+    // should not rely on descendants being initialised yet. This conforms with
+    // native behaviour.
     created.call(this);
-    triggerAttributesCreated(this);
+
+    // We trigger all property handlers for properties that exist once all
+    // descendants are ready for any incoming state updates. This will also
+    // trigger changes for any attributes that are properties.
+    initProps();
+
+    // We trigger all event handlers that have queued up so that nothing has
+    // been lost since they were bound. This is done after initialising
+    // properties because events may depend on initialised values. Any events
+    // triggered from property initialisation will be queued up and initialised
+    // here.
+    initEvents();
```

Comment:
Also, @nicolasv, in light of recent discussion and how this might behave inconsistently is it worth event doing this? To give context, in native, if you put your definitions at the bottom of the page, then descendants will be queued for upgrade before the parent so the parent won't be able to catch any events that are triggered during the `created()` phase of a descendant.